keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,954 @@
+Once the disk is exceeds the limits set by `EvictionSoft` or `EvictionHard`, then that node will eventually be marked as having disk pressure.  
+Garbage collection of containers, images or pods will be kicked off (depending on which filesystem experiences disk pressure).  
+New workloads will not be accepted by that node until disk pressure resolves itself either by garbage collection removing enough or manually intervention.
+
+A mitigation for this is to initially support the case of the writeable layer being on the node filesystem (containerfs = nodefs), so we really are only monitoring two filesystems for pressure.
+
+## Design Details
+
+### CRI
+
+We will switch to using `ImageFsInfo` but this will be guarded by a feature gate.
+
+CRI-O and Containerd return a single element in this case and Kubelet does not assume that there are multiple values in this array.  Regardless, we add an array to ImageFsInfoResponse.
+
+```golang
+// ImageService defines the public APIs for managing images.
+service ImageService {
+…
+    rpc ImageFsInfo(ImageFsInfoRequest) returns (ImageFsInfoResponse) {}
+}
+
+message ImageFsInfoResponse {
+    // Information of image filesystem(s).
+    repeated FilesystemUsage image_filesystems = 1;
+    + // Information of container filesystem(s).
+    + // This is an optional field if container and image
+    + // storage are separated.
+    + // Default will be to return this as empty.
+    + repeated FilesystemUsage container_filesystems = 2;
+}
+```
+
+It is expected of the CRI implementation to return a unique identifier for images and containers so the Kubelet can ask CRI if the objects are stored on separate disks.
+In the dedicated disk for container runtime, images_filesystem and container_fileystem will be set to the same value.
+
+The CRI implementation can set this as needed.  The image and container filesystems are both arrays so this provides some extensibility in case these are stored on multiple disks.  
+
+Container runtimes will need to implement ImageFsInfo
+
+- [CRI-O Implementation](https://github.com/cri-o/cri-o/blob/main/server/image_fs_info.go)
+- [Containerd implementation](https://github.com/containerd/containerd/blob/main/pkg/cri/server/imagefs_info.go)
+
+An Alpha->Beta goal would be to have an implementation of `crictl imagefsinfo` that can allow for more detailed reports of the image fs info.
+
+See [PR](https://github.com/kannon92/cri-tools/pull/1) for an example.
+
+### Stats Summary
+
+Stats Summary has a field called runtime and we will add a containerFs to the runtime field.
+
+```golang
+
+// RuntimeStats are stats pertaining to the underlying container runtime.
+type RuntimeStats struct {
+// Stats about the underlying filesystem where container images are stored.
+// This filesystem could be the same as the primary (root) filesystem.
+// Usage here refers to the total number of bytes occupied by images on the filesystem.
+// +optional
+ImageFs *FsStats `json:"imageFs,omitempty"`
```

Comment:
is it the first occurence where `ImageService` and `RuntimeService` intermingle assuming that one knows about the other?

@mrunalp @haircommander wdyt about this? Would it be a better design to have stats exposed by `ImageService` as well?

Reply:
that's a good point. Currently it's in ImageService, are you suggesting we pull into RuntimeService?

Reply:
I thought the proposal in this KEP to pull stats from RuntiemService? I am thinking if it should move to ImageService. So stats will be pulled with two separate calls

Reply:
There is both a ImageService and a RuntimeService. We pull stats from the image service in the stats providers.

Reply:
sorry, I am slow here. Can you please point which runtime field that RuntimeStats will be added to?